cpu/avr8_common/thread_arch.c

```diff
@@ -281,12 +281,12 @@ void thread_yield_higher(void)
 
 void avr8_exit_isr(void)
 {
-    avr8_state &= ~AVR8_STATE_FLAG_ISR;
+    --avr8_isr_state;
```

Comment:
This is still safe with nested IRQs, right? Because once the ISR resumes, the value of avr8_isr_state will again be the same as when it was interrupted, so interrupted this read-modify-write operation cannot corrupt it.

Maybe add a comment that this doesn't need to be atomic?

Reply:
For ATmega, behavior not change because ISR are disabled. For ATxmega I'm not complete sure and maybe we need ensure an atomic operation.

It was a very good observation @maribu !

Reply:
> This is still safe with nested IRQs, right?

To keep sane, I think it will be necessary a critical section since interrupts are enabled with ATxmega.

I have a doubt about original snip below. Why "reti" only executes after switch context?
It shouldn't be outside if block? It seems odd to me and I need fully understand below to make sure ATxmega will work as expected.

```
void avr8_exit_isr(void)
{
    atmega_state &= ~ATMEGA_STATE_FLAG_ISR;

    /* Force access to avr8_isr_state to take place */
    __asm__ volatile ("" : : : "memory");

    if (sched_context_switch_request) {
        avr8_context_save();
        sched_run();
        avr8_context_restore();
        __asm__ volatile ("reti");
    }
}
```